delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storages/storage/model/Storage.scala

```diff
@@ -0,0 +1,93 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.storages.storage.model
+
+import akka.http.scaladsl.model.Uri
+import ch.epfl.bluebrain.nexus.delta.plugins.storages.storage.contexts
+import ch.epfl.bluebrain.nexus.delta.plugins.storages.storage.model.StorageValue.{DiskStorageValue, RemoteDiskStorageValue, S3StorageValue}
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.context.ContextValue
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.encoder.JsonLdEncoder
+import ch.epfl.bluebrain.nexus.delta.sdk.model.Label
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import io.circe.{Encoder, Json}
+
+sealed trait Storage extends Product with Serializable {
+
+  /**
+    * @return the view id
+    */
+  def id: Iri
+
+  /**
+    * @return a reference to the project that the storage belongs to
+    */
+  def project: ProjectRef
+
+  /**
+    * @return the tag -> rev mapping
+    */
+  def tags: Map[Label, Long]
+
+  /**
+    * @return the original json document provided at creation or update
+    */
+  def source: Json
+
+  /**
+    * @return ''true'' if this store is the project's default, ''false'' otherwise
+    */
+  def default: Boolean
+}
+
+object Storage {
+
+  /**
+    * A storage that stores and fetches files from a local volume
+    */
+  final case class DiskStorage(
+      id: Iri,
+      project: ProjectRef,
+      value: DiskStorageValue,
+      tags: Map[Label, Long],
+      source: Json
+  ) extends Storage {
+    override val default: Boolean = value.default
+  }
+
+  /**
+    * A storage that stores and fetches files from an S3 compatible service
+    */
+  final case class S3Storage(
+      id: Iri,
+      project: ProjectRef,
+      value: S3StorageValue,
+      tags: Map[Label, Long],
+      source: Json
+  ) extends Storage {
+    private[storage] def address(bucket: String): Uri =
+      value.endpoint match {
+        case Some(host) if host.scheme.trim.isEmpty => Uri(s"https://$bucket.$host")
+        case Some(e)                                => e.withHost(s"$bucket.${e.authority.host}")
+        case None                                   => value.region.fold(s"https://$bucket.s3.amazonaws.com")(r => s"https://$bucket.s3.$r.amazonaws.com")
+      }
+    override val default: Boolean                     = value.default
+  }
+
+  /**
+    * A storage that stores and fetches files from a remote volume using a well-defined API
+    */
+  final case class RemoteDiskStorage(
+      id: Iri,
+      project: ProjectRef,
+      value: RemoteDiskStorageValue,
+      tags: Map[Label, Long],
+      source: Json
+  ) extends Storage {
+    override val default: Boolean = value.default
+  }
+
+  val context: ContextValue = ContextValue(contexts.storage)
+
+  implicit private val storageEncoder: Encoder[Storage] = Encoder.instance(_.source)
```

Comment:
if the source tolerates optional fields the source will not be an accurate encoding result